### PR TITLE
Enable user metadata

### PR DESCRIPTION
**Pull request by timiyay**
This PR is a first attempt to add support for Bugsnag's `user` identification feature: https://docs.bugsnag.com/platforms/browsers/#identifying-users

This `user` metadata is a special case that needs to be set on the `Bugsnag` global, not as metadata.

### CODE REVIEW

File: app/instance-initializers/bugsnag.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import Ember  from 'ember';
 import config from '../config/environment';
 import { getContext, generateError } from 'ember-cli-bugsnag/utils/errors';
-import { getMetaData } from '../utils/bugsnag';
+import { getMetaData, getUser } from '../utils/bugsnag';
```
Comment (timiyay) [inline app/instance-initializers/bugsnag.js]: @binhums Wasn't 100% how to structure things here.

If I add a `getUser` method to `app/utils/bugsnag.js`, it works fine. There's a catch - if a user overrides `utils/bugsnag.js`, they _must_ declare both `getMetaData` and `getUser`. If they omit either one, this import will having nothing reference.

Two options off the top of my head:

1. just go with it, and tell users they have to override both methods in the docs
2. split `getMetaData` and `getUser` into `utils/bugsnag-get-metadata.js` and utils/bugsnag-get-user.js` respectively

Thoughts?

Comment (binhums) [inline app/instance-initializers/bugsnag.js]: Agreed that's definitely a problem. I'm relatively offended by the nonstandard filename anyway.

In an ideal world what do you think the API should be for specifying custom methods like `getMetaData` and `getUser`. It seems like an event listener kind of pattern to me.

**Comment by binhums** [comment]
Thanks for putting this together @timiyay  - actually similar to the proposed suggestion in #47 :-).

I don't think the `getMetaData` is the best pattern here. I would rather users either modified the `Bugsnag` instance directly or accessed an API on the addon itself. E.g.

```js

import Bugsnag from 'bugsnag';

Bugsnag.user = {
 ...
};
```

or

```js
import EmberCliBugsnag from 'ember-cli-bugsnag';

EmberCliBugsnag.setUser({
   ...
});
```

What do you think of something like that?

There's some work towards allowing this stuff in #45 but these changes are significant and haven't been majorly tested to so I'm reluctant to release them just yet. On a side note if you're interested in helping me test this in the wild I'd appreciate the help.

**Comment by timiyay** [comment]
@binhums Yes either of these approaches sound good. I will give the `Bugsnag.user` approach a try and, if it works for me, close this.

If that is the best approach - and it sounds good, since it doesn't abstract too far from the original library - then perhaps I'll replace my PR with a docs update instead.

**Comment by binhums** [comment]
I mean I personally like the idea of the user not having to address Bugsnag fundamentals under the hood, especially if it's a relatively common use case I'd like to support it at the high level.

It comes down to a matter of taste and what we're prepared to support. It won't be difficult to implement the second so I might consider doing it myself.

**Comment by binhums** [comment]
And of course, more than happy to pair on a solution.

**Comment by timiyay** [comment]
Sounds like a cool option. I took a quick look at it, though I'm unfamiliar with authoring an Ember CLI add-on.

I tried stashing a function in `index.js`, but it looks like that's not available for import. I guess I'd have to put something in `app/utils`?

**Comment by binhums** [comment]
I found this out the other day, you can export properties / functions through an index file in the addon folder:

https://github.com/ember-cli-reporter/ember-cli-reporter-bugsnag/blob/master/addon/index.js

**Comment by binhums** [comment]
Hey @timiyay are you still interested in putting together a PR? I've made significant changes to master, happy to take you through them.

**Comment by timiyay** [comment]
@binhums yep sure am. I'll take a look at master next time I get a moment, see what's new. Happy to get a quick heads up in this issue, if it's fresh on your brain.

**Comment by binhums** [comment]
It shouldn't change too much. I was thinking something along the lines of:

```js
// ember-cli-bugsnag/addon/index.js

import Bugsnag from 'bugsnag';

export default {
  setUser(user) {
    Bugsnag.user = user;
  }
};
```

**Comment by rogerlevy** [comment]
We're interested in this issue.  Actually I'm still trying to wrap my head around where and how is the best way to pass the user info to bugsnag.  Current plan is to just assign/modify Bugsnag.user at login/session restore.  I don't have an opinion about what is the best way to tell ember-cli-bugsnag.  I have some hours set aside to be involved in this if you're still thinking about it / working on it.

**Comment by binhums** [comment]
@rogerlevy excellent, appreciate any input here. I'm inclined to merge this as is.

If you're interested in pairing on some ember-cli-bugsnag work let me know!

**Comment by timiyay** [comment]
@rogerlevy the solution I ended up with was:

```js
// app/utils/bugsnag.js

/* global Bugsnag */

function setUser(container) {
  const session = container.lookup('service:session');

  if (session.get('isLoggedIn')) {
    const currentUser = session.get('currentUser');

    Bugsnag.user = {
      id: currentUser.get('id'),
      email: currentUser.get('email'),
      name: currentUser.get('name')
    };
  } else {
    Bugsnag.user = {
      id: 'guest'
    };
  }
}

export function getMetaData(_error, container) {
  setUser(container);
  return {};
}
```

This used the existing `getMetaData` hook to call my own function, that modified the `Bugsnag` global directly.

I'll do a quick review of this PR, now that https://github.com/binhums/ember-cli-bugsnag/pull/45 has landed, and get it up to date

**Comment by binhums** [comment]
@timiyay how do you feel about jumping on a call now and getting this cleared up?

**Comment by binhums** [comment]
There is also scope to [https://docs.bugsnag.com/platforms/browsers/#sending-diagnostic-data](do things before each notification). You could try setting this up in an instance-initializer if you wanted. e.g.

```js
app/instance-initializers/set-user-in-bugsnag.js

import Bugsnag from 'bugsnag';

import Ember  from 'ember';
import config from '../config/environment';
import { getContext } from 'ember-cli-bugsnag/utils/errors';
import { getMetaData } from '../utils/bugsnag';
import Bugsnag from 'bugsnag';

var currentEnv = config.environment;

export default {
  name: 'set-user-in-bugsnag',

  initialize: function(instance) {
    Bugsnag.beforeNotify = function(payload, metaData) {
      Bugsnag.user = {
        name: "Long Johnson",
        mood: "Curious"
      };
    });
  };
};
```

**Comment by timiyay** [comment]
@binhums sorry mate I missed [this](https://github.com/binhums/ember-cli-bugsnag/pull/50#issuecomment-250901063) yesterday - it was AFL footy grand final down here.

I like the `beforeNotify` callback, as sticking the the official Bugsnag "way" helps newbies like myself.

The other option I was considering was making 2 optional `utils` methods: `app/utils/bugsnag-get-metadata` and `app/utils/bugsnag-get-user`. This would follow the existing pattern of `getMetadata`

**Comment by timiyay** [comment]
@binhums I've pushed a new commit, which is based off the latest `master`. It separates `getMetadata` and `getUser` into separate utils files called `bugsnag-get-metadata.js` and `bugsnag-get-user.js` respectively.

I went for a documentation-driven approach, meaning that I wanted a custom `getUser` implementation to be a one-liner in the docs.

I still think the `beforeNotify` hook is valuable, and I'll make a separate PR with that. I think that setting a custom user is a common scenario that should have it's own method.

I like the approaches here, as they mirror Bugsnag's own docs. I think we should aim for loose parity with them.

**Comment by binhums** [comment]
OK this looks better, I fundamentally want to remove support for this kind of custom file.

However I'm still not sure how to come up with a way to ensure that the addon is initialized before parts of the app are. Perhaps this is part of the interplay between initilaizers and instance-initializers, this is something I should research.

In the mean time I'm happy to go for this approach. I agree with your todo list. Once these are done I'll merge and then we can start moving on:

- [ ] backwards compatibility with app/utils/bugsnag.js, with an Ember deprecation warning
explain
- [ ] getUser in the docs

Thanks again :heart:

**Comment by seawatts** [comment]
Any progress on this?

**Comment by binhums** [comment]
@seawatts good question. Is this a feature you're interested in? If so let's get this over the line.

File: app/instance-initializers/bugsnag.js [hunk 2]
```diff
@@ -1,7 +1,8 @@
 import Ember  from 'ember';
 import config from '../config/environment';
 import { getContext } from 'ember-cli-bugsnag/utils/errors';
-import { getMetaData } from '../utils/bugsnag';
+import { getMetaData } from '../utils/bugsnag-get-metadata';
```
Comment (patrickberkeley) [inline app/instance-initializers/bugsnag.js]: @binhums would you be opposed to just leaving `/utils/bugsnag` as-is (not renaming) so we can get this merged in?

File: app/instance-initializers/bugsnag.js [hunk 3]
```diff
@@ -22,6 +23,12 @@ export default {
       Ember.onerror = function(error) {
         Bugsnag.context = getContext(router);
         const metaData = getMetaData(error, owner);
+        const user = getUser(owner);
+
+        if (user && typeof user === 'object') {
```
Comment (patrickberkeley) [inline app/instance-initializers/bugsnag.js]: What is the scenario where `user` would not exist or where it wouldn't be an object? `app/utils/bugsnag-get-user.js` always returns and object, correct?

File: app/instance-initializers/bugsnag.js (hunk 2 above)
Comment (binhums) [inline app/instance-initializers/bugsnag.js]: No I wouldn't, let's get this done.

**Comment by binhums** [comment]
@patrickberkeley Are you free to help me test this?

File: app/instance-initializers/bugsnag.js (hunk 3 above)
Comment (binhums) [inline app/instance-initializers/bugsnag.js]: Presumably the only time when this can happen is if the local version of `getUser` returns `null`.

Comment (binhums) [inline app/instance-initializers/bugsnag.js]: I would suggest we change the call to have a default e.g. `const user = getUser(owner) || {};`

**Comment by binhums** [comment]
See #71 for latest developments from @patrickberkeley.

**Comment by binhums** [comment]
Thanks for everyone's work towards this. @patrickberkeley has tidied up the branch and added some tests. Released in [v1.2.0](https://github.com/binhums/ember-cli-bugsnag/releases/tag/v1.2.0)!

**Pull request closed without merging**